src/safeposix/syscalls/net_calls.rs

```diff
@@ -585,47 +640,76 @@ impl Cage {
                             let buflenleft = newbuflen;
                             let mut retval;
 
-                            loop { // we loop here so we can cancel blocking recvs
-                                //socket must be connected so unwrap ok
-                                if sockfdobj.flags & O_NONBLOCK != 0 {
-                                    retval = sockhandle.innersocket.as_ref().unwrap().recvfrom_nonblocking(bufleft, buflenleft, addr);
-                                } else {
-                                    retval = sockhandle.innersocket.as_ref().unwrap().recvfrom(bufleft, buflenleft, addr);
+                            if sockhandle.domain == AF_UNIX {
+                                 // get the remote socket pipe, read from it, and return bytes read
+                                let mut nonblocking = false;
+                                if sockfdobj.flags & O_NONBLOCK != 0 { nonblocking = true;}  
+                                loop {
+                                    let sockinfo = &sockhandle.unix_info.as_ref().unwrap();
+                                    let receivepipe = sockinfo.receivepipe.as_ref().unwrap();
+                                    retval = receivepipe.read_from_pipe(bufleft, buflenleft, nonblocking) as i32;   
+                                    if retval < 0 {
+                                        //If we have already read from a peek but have failed to read more, exit!
+                                        if buflen != buflenleft { return (buflen - buflenleft) as i32; }
+                                        if sockfdobj.flags & O_NONBLOCK == 0 && retval == -(Errno::EAGAIN as i32) {
+                                            if self.cancelstatus.load(interface::RustAtomicOrdering::Relaxed) {
+                                                // if the cancel status is set in the cage, we trap around a cancel point
+                                                // until the individual thread is signaled to cancel itself
+                                                loop{interface::cancelpoint(self.cageid)};
+                                            }
+                                            drop(sockhandle);
+                                            sockhandle = sock_tmp.write();
```

Comment:
If we do not explicitly drop the sockhandle temporarily when looping, it causes a deadlock between `recv` and `send` since both are trying to get write locks on the sockhandle.